float64 floatX
float64 floatY